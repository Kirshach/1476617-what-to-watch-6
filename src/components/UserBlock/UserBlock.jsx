import React from 'react';
import {useSelector} from 'react-redux';
import {Link} from 'react-router-dom';

import {isAuthorizedSelector, userDataSelector} from '../../store/app/auth/selectors.js';
import {AppRoutes} from '../../const.js';

const UserBlock = () => {
  const isAuthorized = useSelector(isAuthorizedSelector);
  const {avatarUrl} = useSelector(userDataSelector);

  return (
    <div className="user-block">
      {isAuthorized ? (
        <div className="user-block__avatar">
          <img
            src={avatarUrl}
            alt="User avatar"
            width={63}
            height={63}
          />
        </div>
      ) : (
        <Link to={AppRoutes.LOGIN} className="user-block__link">
        Sign in
        </Link>
      )}
    </div>
  );
};

export default UserBlock;
