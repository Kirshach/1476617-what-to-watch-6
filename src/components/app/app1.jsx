import React from 'react';
import {useSelector} from 'react-redux';
import {Router, Switch, Route} from 'react-router-dom';

import {isOnlineSelector} from '../../store/app/state/selectors.js';
import {AppRoutes} from '../../const.js';

import OfflineScreen from '../offline-screen/offline-screen.jsx';
import PageNotFound from '../page-not-found/page-not-found';
import PrivateRoute from '../private-route/private-route';
import MoviePlayer from '../movie-player/movie-player';
import AddReview from '../add-review/add-review';
import Logout from '../logout/logout';
import MyList from '../my-list/my-list';
import SignIn from '../sign-in/sign-in';
import Film from '../film/film';
import Main from '../main/main';

import {history} from '../../history';

export const App = () => {
  const isOnline = useSelector(isOnlineSelector);

  return (
    <Router history={history}>
      {
        isOnline ? (
          <Switch>

            <Route exact path={AppRoutes.MAIN}>
              <Main />
            </Route>

            <PrivateRoute exact path={AppRoutes.ADD_REVIEW}>
              <AddReview />
            </PrivateRoute>

            <Route exact path={AppRoutes.FILM}>
              <Film />
            </Route>

            <Route exact path={AppRoutes.LOGIN}>
              <SignIn />
            </Route>

            <Route exact path={AppRoutes.LOGOUT}>
              <Logout/>
            </Route>

            <PrivateRoute exact path={AppRoutes.MY_LIST}>
              <MyList />
            </PrivateRoute>

            <Route exact path={AppRoutes.MOVIE_PLAYER}>
              <MoviePlayer />
            </Route>

            <Route>
              <PageNotFound />
            </Route>


          </Switch>
        ) : (
          <OfflineScreen />
        )
      }
    </Router>
  );
};

App.propTypes = {};

export default App;
