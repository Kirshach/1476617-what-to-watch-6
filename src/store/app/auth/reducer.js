import {ActionType} from './actions';

const HAS_CHECKED_AUTH = `hasCheckedAuth`;
const IS_AUTHORIZED = `isAuthorized`;
const USER_DATA = `userData`;

const initialState = {
  [HAS_CHECKED_AUTH]: false,
  [IS_AUTHORIZED]: false,
  [USER_DATA]: {
    avatarUrl: null,
    email: null,
    name: null,
    id: null,
  },
};

export const authReducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.SET_IS_AUTHORIZED:
      return {
        ...state,
        [IS_AUTHORIZED]: action.payload,
      };
    case ActionType.SET_USER_DATA:
      return {
        ...state,
        [USER_DATA]: action.payload,
      };
    case ActionType.SET_HAS_CHECKED_AUTH:
      return {
        ...state,
        [HAS_CHECKED_AUTH]: action.payload,
      };
    default:
      return state;
  }
};

export const AuthNS = {
  HAS_CHECKED_AUTH,
  IS_AUTHORIZED,
  USER_DATA,
};
