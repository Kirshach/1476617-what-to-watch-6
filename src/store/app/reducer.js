import {ActionType} from './actions';
import {Genres} from '../../const';

export const FILMS_BATCH = 8;

const initialState = {
  filmsShowingCount: FILMS_BATCH,
  genre: Genres.allGenres,
  isLoadingFilms: true,
};

export const appReducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.SET_GENRE:
      return {
        ...state,
        genre: action.payload
      };
    case ActionType.SET_IS_LOADING_FILMS:
      return {
        ...state,
        isLoadingFilms: action.payload
      };
    case ActionType.SHOW_MORE_FILMS:
      return {
        ...state,
        filmsShowingCount: state.filmsShowingCount + (action.payload || FILMS_BATCH),
      };
    case ActionType.SHOW_DEFAULT_FILMS_BATCH:
      return {
        ...state,
        filmsShowingCount: state.filmsShowingCount + (action.payload || FILMS_BATCH),
      };
    default:
      return state;
  }
};
