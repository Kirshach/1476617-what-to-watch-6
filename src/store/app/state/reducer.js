import {Genres} from '../../../const';
import {ActionType} from './actions';

export const FILMS_BATCH = 8;

const REVIEWS_HAVE_LOADED = `reviewsHaveLoaded`;
const FILMS_SHOWING_COUNT = `filmsShowingCount`;
const FILMS_HAVE_LOADED = `filmsHaveLoaded`;
const PROMO_HAS_LOADED = `promoHasLoaded`;
const FILM_HAS_LOADED = `filmHasLoaded`;
const GENRE = `genre`;

const initialState = {
  [REVIEWS_HAVE_LOADED]: false,
  [FILMS_HAVE_LOADED]: false,
  [PROMO_HAS_LOADED]: false,
  [FILM_HAS_LOADED]: false,

  [FILMS_SHOWING_COUNT]: FILMS_BATCH,
  [GENRE]: Genres.allGenres,
};

export const stateReducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.SET_FILM_HAS_LOADED:
      return {
        ...state,
        [FILM_HAS_LOADED]: action.payload,
      };
    case ActionType.SET_FILMS_HAVE_LOADED:
      return {
        ...state,
        [FILMS_HAVE_LOADED]: action.payload,
      };
    case ActionType.SET_PROMO_HAS_LOADED:
      return {
        ...state,
        [PROMO_HAS_LOADED]: action.payload,
      };
    case ActionType.SET_GENRE:
      return {
        ...state,
        [GENRE]: action.payload,
      };
    case ActionType.SET_REVIEWS_HAVE_LOADED:
      return {
        ...state,
        [REVIEWS_HAVE_LOADED]: action.payload,
      };
    case ActionType.SHOW_DEFAULT_FILMS_AMOUNT:
      return {
        ...state,
        [FILMS_SHOWING_COUNT]: FILMS_BATCH,
      };
    case ActionType.SHOW_MORE_FILMS:
      return {
        ...state,
        [FILMS_SHOWING_COUNT]: state[FILMS_SHOWING_COUNT] + (action.payload || FILMS_BATCH),
      };
    default:
      return state;
  }
};

export const StateNS = {
  REVIEWS_HAVE_LOADED,
  FILMS_SHOWING_COUNT,
  FILMS_HAVE_LOADED,
  PROMO_HAS_LOADED,
  FILM_HAS_LOADED,
  GENRE,
};
