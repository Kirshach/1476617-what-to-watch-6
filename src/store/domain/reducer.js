import {ActionType} from './actions';

const FAVOURITE_FILMS = `favouriteFilms`;
const REVIEWS = `reviews`;
const PROMO = `promo`;
const FILMS = `films`;
const FILM = `film`;

const FILM_ID = `id`;
const DATA = `data`;

export const initialState = {
  [FAVOURITE_FILMS]: null,
  [FILMS]: [],
  [FILM]: {},
  [PROMO]: {},
  [REVIEWS]: {
    [FILM_ID]: null,
    [DATA]: [],
  },
};

export const domainReducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.SET_FILM:
      return ({
        ...state,
        [FILM]: action.payload,
      });
    case ActionType.SET_FILMS:
      return ({
        ...state,
        [FILMS]: action.payload,
      });
    case ActionType.SET_FAVOURITE_FILMS:
      return ({
        ...state,
        [FAVOURITE_FILMS]: action.payload,
      });
    case ActionType.SET_PROMO:
      return ({
        ...state,
        [PROMO]: action.payload,
      });
    case ActionType.SET_REVIEWS:
      return ({
        ...state,
        [REVIEWS]: action.payload
      });
    default:
      return state;
  }
};

export const DomainNS = {FILMS, FILM, PROMO, REVIEWS, FAVOURITE_FILMS};
