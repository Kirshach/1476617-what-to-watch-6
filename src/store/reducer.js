import {ActionType} from './actions';
import {Genres} from '../const';
import {films} from '../mocks/films';


const initialState = {
  films,
  genre: Genres.allGenres,
  genres: Object.values(Genres),
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.SET_GENRE:
      return {
        ...state,
        genre: action.payload
      };
    default:
      return state;
  }
};

export {reducer};
