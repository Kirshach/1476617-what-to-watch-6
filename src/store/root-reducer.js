import {combineReducers} from 'redux';

import {appReducer} from './app/reducer';
import {domainReducer} from './domain/reducer';

export const DOMAIN = `domain`;
export const APP = `app`;

const rootReducer = combineReducers({
  [DOMAIN]: domainReducer,
  [APP]: appReducer,
});

export {rootReducer};

export const RootNS = {DOMAIN, APP};
