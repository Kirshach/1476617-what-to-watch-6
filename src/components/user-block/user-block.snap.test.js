import React from 'react';
import {MemoryRouter} from 'react-router-dom';
import {useSelector} from 'react-redux';
import {render} from '@testing-library/react';

import {AuthNS} from '../../store/app/auth/reducer';
import {RootNS} from '../../store/root-reducer';
import {AppNS} from '../../store/app/reducer';
import {store} from '../../mocks/store/store';

import UserBlock from './user-block';

jest.mock(`react-redux`);

describe(`<UserBlock />`, () => {
  test(`renders correctly for an authorized user`, () => {
    useSelector.mockImplementation((selector) => selector(store));

    const {container} = render(
        <MemoryRouter>
          <UserBlock />
        </MemoryRouter>
    );
    expect(container).toMatchSnapshot();
  });

  test(`renders correctly for an unauthorized user`, () => {
    const getUnauthorizedStore = () => {
      const newStore = {...store};
      newStore[RootNS.APP][AppNS.AUTH][AuthNS.IS_AUTHORIZED] = false;
      return newStore;
    };
    useSelector.mockImplementation((selector) => selector(getUnauthorizedStore()));

    const {container} = render(
        <MemoryRouter>
          <UserBlock />
        </MemoryRouter>
    );
    expect(container).toMatchSnapshot();
  });
});
