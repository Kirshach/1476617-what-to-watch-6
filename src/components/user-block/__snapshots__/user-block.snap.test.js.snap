// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserBlock /> renders correctly for an authorized user 1`] = `
<div>
  <div
    class="user-block"
  >
    <a
      class="user-block__link"
      href="/mylist"
    >
      <div
        class="user-block__avatar"
      >
        <img
          alt="User avatar"
          height="63"
          src="https://assets.htmlacademy.ru/intensives/javascript-3/avatar/6.jpg"
          width="63"
        />
      </div>
    </a>
  </div>
</div>
`;

exports[`<UserBlock /> renders correctly for an unauthorized user 1`] = `
<div>
  <div
    class="user-block"
  >
    <a
      class="user-block__link"
      href="/login"
    >
      Sign in
    </a>
  </div>
</div>
`;
